File: src/MartianRobots.Core/Robot.cs
namespace MartianRobots.Core;

public class Robot
{
    public int X { get; }
    public int Y { get; }
    public Orientation Orientation { get; private set; }

    public Robot(int x, int y, Orientation orientation)
    {
        X = x;
        Y = y;
        Orientation = orientation;
    }

    public void TurnLeft()
    {
        Orientation = Orientation switch
        {
            Orientation.North => Orientation.West,
            Orientation.West => Orientation.South,
            Orientation.South => Orientation.East,
            Orientation.East => Orientation.North,
            _ => throw new InvalidOperationException()
        };
    }

    public void TurnRight()
    {
        Orientation = Orientation switch
        {
            Orientation.North => Orientation.East,
            Orientation.East => Orientation.South,
            Orientation.South => Orientation.West,
            Orientation.West => Orientation.North,
            _ => throw new InvalidOperationException()
        };
    }
}

File: go-implementation/robot.go
package main

type Orientation rune

const (
	North Orientation = 'N'
	East  Orientation = 'E'
	South Orientation = 'S'
	West  Orientation = 'W'
)

type Robot struct {
	X           int
	Y           int
	Orientation Orientation
}

func (r *Robot) TurnLeft() {
	switch r.Orientation {
	case North:
		r.Orientation = West
	case West:
		r.Orientation = South
	case South:
		r.Orientation = East
	case East:
		r.Orientation = North
	}
}

func (r *Robot) TurnRight() {
	switch r.Orientation {
	case North:
		r.Orientation = East
	case East:
		r.Orientation = South
	case South:
		r.Orientation = West
	case West:
		r.Orientation = North
	}
}

func (r *Robot) MoveForward() {
	switch r.Orientation {
	case North:
		r.Y++
	case East:
		r.X++
	case South:
		r.Y--
	case West:
		r.X--
	}
}